File: qssbuilder.py
# -*- coding: utf-8 -*-
# /****************************************************************************
# **
# ** Copyright (C) 2014 dragondjf
# **
# ** QFramer is a frame based on Qt5.3, you will be more efficient with it. 
# ** As an Qter，Qt give us a nice coding experience. With user interactive experience(UE) 
# ** become more and more important in modern software, deveployers should consider business and UE.
# ** So, QFramer is born. QFramer's goal is to be a mature solution 
# ** which you only need to be focus on your business but UE for all Qters.
# **
# ** Version  : 0.2.2.0
# ** Author   : dragondjf
# ** Website  : https://github.com/dragondjf
# ** Project  : https://github.com/dragondjf/QCFramer
# ** Blog     : http://my.oschina.net/dragondjf/home/?ft=atme
# ** Wiki     : https://github.com/dragondjf/QCFramer/wiki
# ** Lincence: LGPL V2
# ** QQ: 465398889
# ** Email: dragondjf@gmail.com, ding465398889@163.com, 465398889@qq.com
# ** 
# ****************************************************************************/
import os
import sys
import copy

qssFolder = os.sep.join([os.getcwd(), 'skin', 'qss'])
qssTemplatePath = os.sep.join([qssFolder, 'template.qss'])


theme_green = {
    'main_background_color': "qlineargradient(spread:pad, x1:0, y1:1, x2:0, y2:0, stop:0 rgba(11, 158, 12, 255), stop:0.5 rgba(150, 220, 120, 255), stop:1 rgba(11, 158, 12, 255))",   # 主窗体背景颜色
    'main_background_image': "none",
    'nav_background_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:0, stop:0 rgba(11, 158, 12, 255), stop:1 rgba(150, 220, 120, 255))", # 导航条背景颜色
    'main_hover_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:1, stop:0 rgba(11, 158, 12, 255), stop:0.5 rgba(150, 220, 120, 255), stop:1 rgba(11, 158, 12, 255))", # 导航条hover背景颜色
    'main_select_color': "rgba(11, 158, 12, 255)", # 导航条select背景颜色
    'statusBar_background_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:1, stop:0 rgba(11, 158, 12, 255), stop:0.5 rgba(150, 220, 120, 255), stop:1 rgba(11, 158, 12, 255))",
    'border_separator': "0px solid rgb(0, 206, 161)", # 分割线背景
    'Ftitle_color': 'white', # 标题title字体颜色
    'Menu_border_left': "5px solid rgb(255, 127, 39)", # 菜单左边界
    'AntimationLine_background-color': 'green', # 页面切换时动画进度条的背景颜色
    'Label_color': "white", # label字体颜色
    'PushButton_color': "white", # 按钮字体颜色
    'PushButton_background_color': "rgb(88, 197, 1)", #按钮背景颜色
    'PushButton_Pressed_background_color': "rgb(76, 148, 31)",
    'PushButton_border_left': "5px solid rgb(255, 127, 39)", #按钮左边界颜色
    'PushButton_border_bottom': "0px solid rgb(255, 127, 39)", #按钮下边界颜色
    'Combox_background_color': "rgb(0, 0, 64)", #combox背景颜色
    'ScrollBar_background_color': "white", #滚动条背景颜色
    'ScrollBar_handle_color': "lightgray", #滚动条滚动颜色
    'HeadView_background_color': "rgb(21, 206, 109)", # headview背景颜色
    'TableTree_background_color': "rgb(88, 197, 1)" # tabtree背景颜色
}


theme_blue = {
    'main_background_color': "rgb(51, 79, 91)",   # 主窗体背景颜色
    'main_background_image': "none",
    'nav_background_color': "rgb(43, 152, 188)", # 导航条背景颜色
    'main_hover_color': "rgb(27, 118, 150)", # 导航条hover背景颜色
    'main_select_color': "rgb(34, 125, 155)", # 导航条select背景颜色
    'border_separator': "1px solid rgb(34, 125, 155)", # 分割线背景
    'Ftitle_color': 'white', # 标题title字体颜色
    'Menu_border_left': "5px solid rgb(255, 127, 39)", # 菜单左边界
    'AntimationLine_background-color': 'rgb(255, 127, 39)', # 页面切换时动画进度条的背景颜色
    'Label_color': "rgb(53, 153, 135)", # label字体颜色
    'PushButton_color': "white", # 按钮字体颜色
    'PushButton_background_color': "rgb(43, 65, 77)", #按钮背景颜色
    'PushButton_Pressed_background_color': "rgb(50, 77, 92)",
    'PushButton_border_left': "5px solid rgb(255, 127, 39)", #按钮左边界颜色
    'PushButton_border_bottom': "0px solid rgb(255, 127, 39)", #按钮下边界颜色
    'Combox_background_color': "rgb(0, 0, 64)", #combox背景颜色
    'ScrollBar_background_color': "white", #滚动条背景颜色
    'ScrollBar_handle_color': "rgb(70, 97, 102)", #滚动条滚动颜色
    'HeadView_background_color': "rgb(34, 125, 155)", # headview背景颜色
    'TableTree_background_color': "rgb(27, 118, 150)" # tabtree背景颜色
}


theme_gradient= {
    'main_background_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:1, stop:0 rgba(0, 82, 112, 255), stop:0.5 rgba(0, 211, 197, 255), stop:1 rgba(0, 82, 112, 255))",   # 主窗体背景颜色
    'main_background_image': "none",
    'nav_background_color': "qlineargradient(spread:pad, x1:-1, y1:0, x2:1 y2:0, stop:0 rgba(0, 211, 197, 255), stop:1 rgba(0, 82, 112, 255))", # 导航条背景颜色
    'main_hover_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:0, stop:0 rgba(0, 82, 112, 255), stop:0.5 rgba(0, 211, 197, 255), stop:1 rgba(0, 82, 112, 255))", # 导航条hover背景颜色
    'main_select_color': "rgb(0, 82, 112)", # 导航条select背景颜色
    'statusBar_background_color': "qlineargradient(spread:pad, x1:0, y1:0, x2:1, y2:0, stop:0 rgba(0, 82, 112, 255), stop:0.5 rgba(0, 211, 197, 255), stop:1 rgba(0, 82, 112, 255))",
    'border_separator': "1px solid rgb(0, 82, 112)", # 分割线背景
    'Ftitle_color': 'white', # 标题title字体颜色
    'Menu_border_left': "5px solid rgba(0, 211, 197, 255)", # 菜单左边界
    'AntimationLine_background-color': 'rgba(0, 211, 197, 255)', # 页面切换时动画进度条的背景颜色
    'Label_color': "white", # label字体颜色
    'PushButton_color': "white", # 按钮字体颜色
    'PushButton_background_color': "rgb(43, 65, 77)", #按钮背景颜色
    'PushButton_Pressed_background_color': "rgb(50, 77, 92)",
    'PushButton_border_left': "5px solid rgba(0, 211, 197, 255)", #按钮左边界颜色
    'PushButton_border_bottom': "0px solid rgb(255, 127, 39)", #按钮下边界颜色
    'Combox_background_color': "rgb(0, 0, 64)", #combox背景颜色
    'ScrollBar_background_color': "white", #滚动条背景颜色
    'ScrollBar_handle_color': "rgb(70, 97, 102)", #滚动条滚动颜色
    'HeadView_background_color': "rgb(0, 82, 112)", # headview背景颜色
    'TableTree_background_color': "rgb(27, 118, 150)" # tabtree背景颜色
}



theme = theme_green

if __name__ == '__main__':
    if len(sys.argv) >= 2:
        newQssFileName = sys.argv[1]
    else:
        newQssFileName = "test.qss"

    qssOutFileName = os.sep.join([qssFolder, newQssFileName])

    with open(qssTemplatePath) as f:
        qssContent = f.read()
        for key, value in theme.items():
            qssContent = qssContent.replace(key, value)

    with open(qssOutFileName, 'w') as f:
        f.write(qssContent)

File: functionpages/qssbuilder.cpp
/****************************************************************************
**
** Copyright (C) 2014 dragondjf
**
** QFramer is a frame based on Qt5.3, you will be more efficient with it. 
** As an Qter, Qt give us a nice coding experience. With user interactive experience(UE) 
** become more and more important in modern software, deveployers should consider business and UE.
** So, QFramer is born. QFramer's goal is to be a mature solution 
** which you only need to be focus on your business but UE for all Qters.
**
** Version	: 0.2.5.0
** Author	: dragondjf
** Website	: https://github.com/dragondjf
** Project	: https://github.com/dragondjf/QCFramer
** Blog		: http://my.oschina.net/dragondjf/home/?ft=atme
** Wiki		: https://github.com/dragondjf/QCFramer/wiki
** Lincence: LGPL V2
** QQ: 465398889
** Email: dragondjf@gmail.com, ding465398889@163.com, 465398889@qq.com
** 
****************************************************************************/

#include "qssbuilder.h"
#include "QFramer/futil.h"
#include <QGridLayout>
#include <QColorDialog>
#include <QLabel>
#include <QFile>
#include <QDir>
#include <QHBoxLayout>
#include <QRadioButton>
#include <QButtonGroup>

#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0)
    #include <QJsonDocument>
    #include <QJsonObject>
#endif


QssBuilder::QssBuilder(QWidget *parent) :
    QFrame(parent)
{
    initData();
    initUI();
    initConnect();
}

void QssBuilder::initData()
{
    colorKeys << "main background color";
    colorLabels << "main_background_color";
    colorValues << "rgb(51, 79, 91)";

    colorKeys << "main background image";
    colorLabels << "main_background_image";
    colorValues << "none";

    colorKeys << "navgation background color";
    colorLabels << "nav_background_color";
    colorValues << "rgb(43, 152, 188)";

    colorKeys << "main hover color";
    colorLabels << "main_hover_color";
    colorValues << "rgb(27, 118, 150)";

    colorKeys << "main select color";
    colorLabels << "main_select_color";
    colorValues << "rgb(34, 125, 155)";

    colorKeys << "statusBar background color";
    colorLabels << "statusBar_background_color";
    colorValues << "rgb(34, 125, 155)";

    colorKeys << "separator border";
    colorLabels << "border_separator";
    colorValues << "1px solid rgb(34, 125, 155)";

    colorKeys << "title color";
    colorLabels << "Ftitle_color";
    colorValues << "white";

    colorKeys << "menu left border";
    colorLabels << "Menu_border_left";
    colorValues << "5px solid rgb(255, 127, 39)";

    colorKeys << "animation line background color";
    colorLabels << "AntimationLine_background-color";
    colorValues << "rgb(255, 127, 39)";

    colorKeys << "label font color";
    colorLabels << "Label_color";
    colorValues << "rgb(53, 153, 135)";

    colorKeys << "pusbutton color";
    colorLabels << "PushButton_color";
    colorValues << "white";

    colorKeys << "pusbutton background color";
    colorLabels << "PushButton_background_color";
    colorValues << "rgb(43, 65, 77)";

    colorKeys << "pusbutton pressed background color";
    colorLabels << "PushButton_Pressed_background_color";
    colorValues << "rgb(50, 77, 92)";

    colorKeys << "pusbutton left border";
    colorLabels << "PushButton_border_left";
    colorValues << "5px solid rgb(255, 127, 39)";

    colorKeys << "pusbutton bottom border";
    colorLabels << "PushButton_border_bottom";
    colorValues << "0px solid rgb(255, 127, 39)";

    colorKeys << "combox background color";
    colorLabels << "Combox_background_color";
    colorValues << "rgb(0, 0, 64)";

    colorKeys << "scrollbar background color";
    colorLabels << "ScrollBar_background_color";
    colorValues << "white";

    colorKeys << "scrollbar handle color";
    colorLabels << "ScrollBar_handle_color";
    colorValues << "rgb(70, 97, 102)";

    colorKeys << "headview background color";
    colorLabels << "HeadView_background_color";
    colorValues << "rgb(34, 125, 155)";

    colorKeys << "table or tree background color";
    colorLabels << "TableTree_background_color";
    colorValues << "rgb(27, 118, 150)";
}

void QssBuilder::initUI()
{
    QGridLayout* themeLayout = new QGridLayout;
    for(int i=0; i< colorKeys.length(); i++)
    {
        QLabel* label = new QLabel("<h4>" + colorKeys[i] + ":" + "</h4>");
        label->setObjectName("qssBuilderLabel");
        label->setFixedSize(280, 24);
        QLineEdit* lineedit = new QLineEdit(colorValues[i]);
        lineedit->setFixedSize(300, 24);
        QPushButton* button = new QPushButton;

        QString style = QString("QPushButton#setColor{background-color: %1;}");
        style = style.arg(colorValues[i]);
        button->setStyleSheet(style);
        button->setObjectName("setColor");
        button->setFixedSize(60, 24);
        lineedits.append(lineedit);
        buttons.append(button);

        themeLayout->addWidget(label, i, 0);
        themeLayout->addWidget(lineedit, i, 1);
        themeLayout->addWidget(button, i, 2);
        themeLayout->setColumnStretch(4, 1);

        colorMap[colorLabels[i]] = colorValues[i];
    }


    QLabel* iconLabel = new QLabel(tr("Icon color:"));
    QRadioButton* blackButton = new QRadioButton(tr("Balck"));
    QRadioButton* whiteButton = new QRadioButton(tr("White"));
    whiteButton->setChecked(true);
    QButtonGroup* iconGroup = new QButtonGroup;
    iconGroup->addButton(blackButton, 0);
    iconGroup->addButton(whiteButton, 1);
    connect(iconGroup, SIGNAL(buttonClicked(int)), this, SLOT(changeIconColor(int)));

    QGridLayout* previwerLayout = new QGridLayout;
    previwerLayout->addWidget(iconLabel, 0, 0);
    previwerLayout->addWidget(blackButton, 0, 1);
    previwerLayout->addWidget(whiteButton, 0, 2);
    previwerLayout->setColumnStretch(4, 1);
    previwerLayout->setRowStretch(1, 1);
    QHBoxLayout* mainLayout = new QHBoxLayout;
    mainLayout->addLayout(themeLayout);
//    mainLayout->addLayout(previwerLayout);
    setLayout(mainLayout);
}

void QssBuilder::initConnect()
{
    for(int i=0; i<buttons.length(); i++){
        connect(buttons[i], SIGNAL(clicked()), this, SLOT(getColor()));
        connect(lineedits[i], SIGNAL(textChanged(QString)), this, SLOT(setButtonColor(QString)));
    }
}

void QssBuilder::getColor()
{
    currentIndex = buttons.indexOf((QPushButton*)sender());
    QColorDialog* colorDialog = new QColorDialog();
    colorDialog->setObjectName("colorDialog");
    connect(colorDialog, SIGNAL(currentColorChanged(QColor)), this, SLOT(setButtonColor(QColor)));
    colorDialog->exec();
}

void QssBuilder::setButtonColor(QColor color)
{
    QString ctr = QString("rgb(%1, %2, %3)");
    ctr = ctr.arg(QString::number(color.red()), QString::number(color.green()), QString::number(color.blue()));
    QString style = QString("QPushButton#setColor{background-color: %1;}");
    style = style.arg(ctr);
    lineedits.at(currentIndex)->setText(ctr);
      buttons.at(currentIndex)->setStyleSheet(style);

    colorMap[colorLabels[currentIndex]] = ctr;
    updateTheme();
}

void QssBuilder::setButtonColor(QString color)
{
    currentIndex = lineedits.indexOf((QLineEdit*)sender());
    colorMap[colorLabels[currentIndex]] = color;
    QString style = QString("QPushButton#setColor{background-color: %1;}");
    style = style.arg(color);
    buttons.at(currentIndex)->setStyleSheet(style);
    updateTheme();
}

void QssBuilder::updateTheme()
{

//#if QT_VERSION >= QT_VERSION_CHECK(5, 0, 0)
//    QJsonDocument doc = QJsonDocument(QJsonObject::fromVariantMap(colorMap));
//    QFile outFile("customColor.json");
//    outFile.open(QIODevice::WriteOnly);
//    outFile.write(doc.toJson());
//    outFile.close();
//#endif

    QString qss = getQssFromFile(":/qss/skin/qss/template.qss");
    for(int i=0; i< colorLabels.length(); i++){
       qss.replace(colorLabels[i], colorMap[colorLabels[i]]);
    }
    qApp->setStyleSheet(qss);

    QFile outqssFile("test.qss");
    outqssFile.open(QIODevice::WriteOnly);
    outqssFile.write(qss.toStdString().data());
    outqssFile.close();
}


void QssBuilder::changeIconColor(int mode)
{
    QString qss = getQssFromFile(":/qss/skin/qss/template.qss");
    for(int i=0; i< colorLabels.length(); i++){
       qss.replace(colorLabels[i], colorMap[colorLabels[i]]);
    }

    if(mode == 0){
        qss.replace(QString(":/skin/icons/dark/"), QString(":/skin/icons/light/"));
    }else if(mode == 1){
        qss.replace(QString(":/skin/icons/light/"), QString(":/skin/icons/dark/"));
    }

    qApp->setStyleSheet(qss);
    QFile outqssFile("test.qss");
    outqssFile.open(QIODevice::WriteOnly);
    outqssFile.write(qss.toStdString().data());
    outqssFile.close();
}
